### PR TITLE
- match definition of parity function to sgn0 of ietf draft in swu map

### PR DESCRIPTION
**Pull request by drskalman**
- adapt unit tests to the new definition

## Description

closes: https://github.com/arkworks-rs/algebra/pull/343#discussion_r781707918

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Review by InaOana — COMMENTED** [review]
From my side, code and tests match the sig0 ieft draft.

File: ec/src/hashing/curve_maps/swu/mod.rs [hunk 1]
```diff
@@ -188,34 +191,45 @@ mod tests {
...
+
         let element_test1 = Fq6::new(a1, a2, a3);
         let element_test2 = Fq6::new(a2, a3, a4);
         let element_test3 = Fq6::new(a3, a4, a1);
         let element_test4 = Fq6::new(a4, a1, a2);
-        assert_eq!(Fq6::parity(&element_test1), false);
-        assert_eq!(Fq6::parity(&element_test2), false);
+        let element_test5 = Fq6::new(a1, a5, a2);
+
+        assert_eq!(Fq6::parity(&element_test1), true, "parity is the oddness of first none-zero coefficient of element represented over the prime field");
```
Comment (mmagician) [inline ec/src/hashing/curve_maps/swu/mod.rs:229]:
```suggestion
        assert_eq!(Fq6::parity(&element_test1), true, "parity is the oddness of first non-zero coefficient of element represented over the prime field");
```

File: ec/src/hashing/curve_maps/swu/mod.rs [hunk 2]
```diff
@@ -188,34 +191,45 @@ mod tests {
...
+        assert_eq!(Fq6::parity(&element_test2), true, "parity is the oddness of first none-zero coefficient of element represented over the prime field");
```
Comment (mmagician) [inline ec/src/hashing/curve_maps/swu/mod.rs:230]:
```suggestion
        assert_eq!(Fq6::parity(&element_test2), true, "parity is the oddness of first non-zero coefficient of element represented over the prime field");
```

File: ec/src/hashing/curve_maps/swu/mod.rs [hunk 3]
```diff
@@ -188,34 +191,45 @@ mod tests {
...
+        assert_eq!(Fq::parity(&a3), false);
+    }
+
+    #[test]
+    fn test_parity_of_quadratic_extension_elements() {
+        let element_test1 = Fq2::new(Fq::from(0), Fq::from(1));
+        let element_test2 = Fq2::new(Fq::from(1), Fq::from(0));
+        let element_test3 = Fq2::new(Fq::from(10), Fq::from(5));
+        let element_test4 = Fq2::new(Fq::from(5), Fq::from(10));
+        assert_eq!(Fq2::parity(&element_test1), true, "parity is the oddness of first none-zero coefficient of element represented over the prime field" );
```
Comment (mmagician) [inline ec/src/hashing/curve_maps/swu/mod.rs:209]:
```suggestion
        assert_eq!(Fq2::parity(&element_test1), true, "parity is the oddness of first non-zero coefficient of element represented over the prime field" );
```

**Review by mmagician — CHANGES_REQUESTED** [review]

File: ec/src/hashing/curve_maps/swu/mod.rs [hunk 4]
```diff
@@ -30,15 +30,18 @@ pub struct SWUMap<P: SWUParams> {
...
             .to_base_prime_field_elements()
-            .next()
-            .unwrap()
-            .into_repr()
-            .is_odd()
+            .find(|&x| !x.is_zero())
+        {
+            Some(x) => x.into_repr().is_odd(),
+            _ => false,
+        }
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/swu/mod.rs:44]:
```suggestion
            .find(|&x| !x.is_zero())
            .map_or(false, |x| x.into_repr().is_odd())
```

File: ec/src/hashing/curve_maps/swu/mod.rs [hunk 5]
```diff
@@ -30,15 +30,18 @@ pub struct SWUMap<P: SWUParams> {
...
 }
 
-/// Trait defining a parity method on the Field elements.
+/// Trait defining a parity method on the Field elements based on [\[1\]] Section 4.1
+///
+/// - [\[1\]] <https://datatracker.ietf.org/doc/draft-irtf-cfrg-hash-to-curve/>
 trait ElementParity<F: Field> {
     fn parity(element: &F) -> bool {
-        element
+        match element
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/swu/mod.rs:38]:
```suggestion
        element
```

**Review by Pratyush — APPROVED** [review]

**Review by mmagician — APPROVED** [review]

**Pull request merged**